File: lib/dotsync/models/mapping.rb
# frozen_string_literal: true

module Dotsync
  # Mapping represents a source-to-destination path pair for synchronization.
  #
  # A mapping defines what should be synced (src -> dest), with optional filters:
  # - `only`: whitelist of paths to include (everything else is excluded)
  # - `ignore`: blacklist of paths to exclude
  # - `force`: enable removal detection (find dest files not in src)
  #
  # == Path Matching Methods
  #
  # The class provides several methods for path filtering, each with specific use cases:
  #
  # - #include?(path): Returns true if path is inside an inclusion. Used for file filtering.
  # - #bidirectional_include?(path): Returns true if path is inside OR contains an inclusion.
  #   Used during directory traversal to allow descending into parent directories.
  # - #should_prune_directory?(path): Returns true if a directory subtree can be skipped entirely.
  #   This is a PERFORMANCE OPTIMIZATION for Find.prune - see DirectoryDiffer for details.
  #
  class Mapping
    include Dotsync::PathUtils

    attr_reader :original_src, :original_dest, :original_ignores, :original_only

    def initialize(attributes)
      @original_src = attributes["src"]
      @original_dest = attributes["dest"]
      @original_ignores = Array(attributes["ignore"])
      @original_only = Array(attributes["only"])
      @force = attributes["force"] || false
      @hooks = Array(attributes["hooks"])

      @sanitized_src, @sanitized_dest, @sanitized_ignores, @sanitized_only = process_paths(
        @original_src,
        @original_dest,
        @original_ignores,
        @original_only
      )
    end

    def src
      @sanitized_src
    end

    def dest
      @sanitized_dest
    end

    def ignores
      @sanitized_ignores
    end

    def inclusions
      @sanitized_only
    end

    def force?
      @force
    end

    attr_reader :hooks

    def has_hooks?
      @hooks.any?
    end

    def directories?
      File.directory?(src) && File.directory?(dest)
    end

    def files?
      files_present? || file_present_in_src_only?
    end

    def files_present?
      File.file?(src) && File.file?(dest)
    end

    def file_present_in_src_only?
      File.file?(src) && !File.exist?(dest) && File.directory?(File.dirname(dest))
    end

    def valid?
      return false unless paths_are_distinct?
      return false unless paths_not_nested?
      directories? || files? || file_present_in_src_only?
    end

    def file_changed?
      return false unless files_present?
      # Check size first for quick comparison
      return true if File.size(src) != File.size(dest)
      # If sizes match, compare content
      FileUtils.compare_file(src, dest) == false
    end

    def backup_possible?
      valid? && File.exist?(dest)
    end

    def backup_basename
      return unless valid?
      return File.dirname(dest) unless File.exist?(dest)
      File.basename(dest)
    end

    def icons
      msg = []
      msg << Icons.force if force?
      msg << Icons.only if has_inclusions?
      msg << Icons.ignore if has_ignores?
      msg << Icons.hook if has_hooks?
      msg << Icons.invalid unless valid?
      msg.join
    end

    def to_s
      msg = "#{decorated_src} → #{decorated_dest}"
      msg += " #{icons}" if icons != ""
      msg
    end

    def decorated_src
      colorize_env_vars(original_src)
    end

    def decorated_dest
      colorize_env_vars(original_dest)
    end

    def apply_to(path)
      relative_path = if Pathname.new(path).absolute?
        path.delete_prefix(File.join(src, "/"))
      else
        path
      end

      Dotsync::Mapping.new(
        "src" => File.join(@original_src, relative_path),
        "dest" => File.join(@original_dest, relative_path),
        "force" => @force,
        "only" => @only,
        "ignore" => @original_ignores
      )
    end

    def include?(path)
      return true unless has_inclusions?
      return true if path == src
      inclusions.any? { |inclusion| path_is_parent_or_same?(inclusion, path) }
    end

    def bidirectional_include?(path)
      return true unless has_inclusions?
      return true if path == src
      inclusions.any? { |inclusion| path_is_parent_or_same?(inclusion, path) || path_is_parent_or_same?(path, inclusion) }
    end

    def ignore?(path)
      ignores.any? { |ignore| path.start_with?(ignore) }
    end

    def skip?(path)
      ignore?(path) || !include?(path)
    end

    # Determines if a directory subtree can be entirely skipped during traversal.
    #
    # PERFORMANCE OPTIMIZATION: This method enables Find.prune in DirectoryDiffer.
    # When walking large destination directories (e.g., ~/.config with 8,686 files),
    # pruning irrelevant subtrees avoids visiting thousands of files that will never
    # match the `only` filter. This reduced scan time from 7.2s to 0.5s in benchmarks.
    #
    # A directory should be pruned if:
    # 1. It's ignored (in the ignore list), OR
    # 2. It has inclusions AND the path is neither:
    #    - Inside an inclusion (would be synced)
    #    - A parent of an inclusion (might contain synced files)
    #
    # @param path [String] Absolute path to check
    # @return [Boolean] true if the entire directory subtree can be skipped
    def should_prune_directory?(path)
      return true if ignore?(path)
      return false unless has_inclusions?
      !bidirectional_include?(path)
    end

    private
      def has_ignores?
        @original_ignores.any?
      end

      def has_inclusions?
        @original_only.any?
      end

      def process_paths(raw_src, raw_dest, raw_ignores, raw_only)
        sanitized_src = sanitize_path(raw_src)
        sanitized_dest = sanitize_path(raw_dest)
        sanitized_ignores = raw_ignores.flat_map do |path|
          [File.join(sanitized_src, path), File.join(sanitized_dest, path)]
        end
        sanitized_only = raw_only.flat_map do |path|
          [File.join(sanitized_src, path), File.join(sanitized_dest, path)]
        end
        [sanitized_src, sanitized_dest, sanitized_ignores, sanitized_only]
      end

      def paths_are_distinct?
        src != dest
      end

      def paths_not_nested?
        # Check if dest is inside src or vice versa
        return false if dest.start_with?("#{src}/")
        return false if src.start_with?("#{dest}/")
        true
      end
  end
end

File: internal/model/mapping.go
// Package model holds dotsync's core domain types. Mapping is a faithful port
// of Ruby's Dotsync::Mapping: a source-to-destination path pair with optional
// force/only/ignore/hook behavior, plus the path-matching predicates the diff
// engine relies on. Method semantics mirror the Ruby original exactly; where a
// Ruby quirk is load-bearing (e.g. two different ignore semantics), it is
// reproduced and noted.
package model

import (
	"os"
	"path/filepath"
	"strings"

	"github.com/dsaenztagarro/dotsync/internal/fsutil"
	"github.com/dsaenztagarro/dotsync/internal/paths"
)

// InvalidityReason enumerates why a mapping is not valid. The empty string
// means valid (Ruby returns nil).
type InvalidityReason string

const (
	Valid       InvalidityReason = ""
	PathsSame   InvalidityReason = "paths_same"
	PathsNested InvalidityReason = "paths_nested"
	SrcMissing  InvalidityReason = "src_missing"
	DestMissing InvalidityReason = "dest_missing"
)

// shorthandLocalBases maps a sync shorthand to its local (non-mirror) base
// path expression, used by ManifestKey. Mirrors
// SyncMappings::SHORTHANDS[type][:local].
var shorthandLocalBases = map[string]string{
	"home":       "$HOME",
	"xdg_config": "$XDG_CONFIG_HOME",
	"xdg_data":   "$XDG_DATA_HOME",
	"xdg_cache":  "$XDG_CACHE_HOME",
	"xdg_bin":    "$XDG_BIN_HOME",
}

// Attributes is the normalized input for constructing a Mapping. The config
// layer performs the Ruby `Array(...)` coercion (string-or-array) before
// building this struct.
type Attributes struct {
	Src      string
	Dest     string
	Ignore   []string
	Only     []string
	Force    bool
	Hooks    []string
	SyncType string
}

// Mapping is a single src -> dest synchronization unit.
type Mapping struct {
	originalSrc     string
	originalDest    string
	originalIgnores []string
	originalOnly    []string
	force           bool
	hooks           []string
	syncType        string

	sanitizedSrc     string
	sanitizedDest    string
	sanitizedIgnores []string
	sanitizedOnly    []string
}

// New builds a Mapping, sanitizing paths the way Ruby's initialize does via
// process_paths (each ignore/only entry expands to BOTH src- and dest-relative
// absolute forms).
func New(attrs Attributes) *Mapping {
	m := &Mapping{
		originalSrc:     attrs.Src,
		originalDest:    attrs.Dest,
		originalIgnores: append([]string(nil), attrs.Ignore...),
		originalOnly:    append([]string(nil), attrs.Only...),
		force:           attrs.Force,
		hooks:           append([]string(nil), attrs.Hooks...),
		syncType:        attrs.SyncType,
	}
	m.sanitizedSrc = paths.SanitizePath(attrs.Src)
	m.sanitizedDest = paths.SanitizePath(attrs.Dest)
	m.sanitizedIgnores = expandFilters(m.sanitizedSrc, m.sanitizedDest, attrs.Ignore)
	m.sanitizedOnly = expandFilters(m.sanitizedSrc, m.sanitizedDest, attrs.Only)
	return m
}

// expandFilters joins each entry onto both the sanitized src and dest, matching
// process_paths' flat_map.
func expandFilters(src, dest string, entries []string) []string {
	out := make([]string, 0, len(entries)*2)
	for _, e := range entries {
		out = append(out, filepath.Join(src, e), filepath.Join(dest, e))
	}
	return out
}

// --- accessors (sanitized) ---

func (m *Mapping) Src() string          { return m.sanitizedSrc }
func (m *Mapping) Dest() string         { return m.sanitizedDest }
func (m *Mapping) Ignores() []string    { return m.sanitizedIgnores }
func (m *Mapping) Inclusions() []string { return m.sanitizedOnly }
func (m *Mapping) Force() bool          { return m.force }
func (m *Mapping) Hooks() []string      { return m.hooks }
func (m *Mapping) SyncType() string     { return m.syncType }

// --- accessors (original / unexpanded) ---

func (m *Mapping) OriginalSrc() string       { return m.originalSrc }
func (m *Mapping) OriginalDest() string      { return m.originalDest }
func (m *Mapping) OriginalIgnores() []string { return m.originalIgnores }

func (m *Mapping) HasHooks() bool      { return len(m.hooks) > 0 }
func (m *Mapping) HasInclusions() bool { return len(m.originalOnly) > 0 }
func (m *Mapping) HasIgnores() bool    { return len(m.originalIgnores) > 0 }

// ManifestKey returns the stable orphan-manifest key for a shorthand mapping,
// or "" when the mapping has no sync_type. Mirrors Mapping#manifest_key.
func (m *Mapping) ManifestKey() string {
	if m.syncType == "" {
		return ""
	}
	base, ok := shorthandLocalBases[m.syncType]
	if !ok {
		return m.syncType
	}
	expandedBase := paths.SanitizePath(base)
	switch {
	case m.sanitizedDest == expandedBase:
		return m.syncType
	case strings.HasPrefix(m.sanitizedDest, expandedBase+"/"):
		subpath := strings.TrimPrefix(m.sanitizedDest, expandedBase+"/")
		return m.syncType + "--" + subpath
	default:
		return m.syncType
	}
}

// --- type predicates (touch the filesystem, following symlinks like Ruby) ---

func (m *Mapping) Directories() bool { return isDir(m.sanitizedSrc) && isDir(m.sanitizedDest) }

func (m *Mapping) Files() bool { return m.FilesPresent() || m.FilePresentInSrcOnly() }

func (m *Mapping) FilesPresent() bool { return isFile(m.sanitizedSrc) && isFile(m.sanitizedDest) }

// FilePresentInSrcOnly reports src is a file, dest is missing, but dest's
// parent directory exists.
func (m *Mapping) FilePresentInSrcOnly() bool {
	return isFile(m.sanitizedSrc) && !exists(m.sanitizedDest) && isDir(filepath.Dir(m.sanitizedDest))
}

// --- validity ---

func (m *Mapping) Valid() bool { return m.InvalidityReason() == Valid }

// InvalidityReason mirrors the Ruby ordering: distinctness and nesting are
// checked before existence.
func (m *Mapping) InvalidityReason() InvalidityReason {
	if m.sanitizedSrc == m.sanitizedDest {
		return PathsSame
	}
	if !m.pathsNotNested() {
		return PathsNested
	}
	if m.Directories() || m.Files() {
		return Valid
	}
	if !exists(m.sanitizedSrc) {
		return SrcMissing
	}
	return DestMissing
}

func (m *Mapping) pathsNotNested() bool {
	if strings.HasPrefix(m.sanitizedDest, m.sanitizedSrc+"/") {
		return false
	}
	if strings.HasPrefix(m.sanitizedSrc, m.sanitizedDest+"/") {
		return false
	}
	return true
}

// DestCreatable reports whether the only problem is a missing destination that
// mkdir can fix.
func (m *Mapping) DestCreatable() bool {
	return m.InvalidityReason() == DestMissing && m.fixableByMkdir()
}

func (m *Mapping) fixableByMkdir() bool {
	switch {
	case isDir(m.sanitizedSrc):
		return !exists(m.sanitizedDest)
	case isFile(m.sanitizedSrc):
		return !exists(m.sanitizedDest) && !exists(filepath.Dir(m.sanitizedDest))
	default:
		return false
	}
}

// CreateDest creates the destination directory (or the parent directory for a
// file dest). Mirrors create_dest!.
func (m *Mapping) CreateDest() error {
	target := m.sanitizedDest
	if !isDir(m.sanitizedSrc) {
		target = filepath.Dir(m.sanitizedDest)
	}
	return os.MkdirAll(target, 0o755)
}

// FileChanged reports whether a single-file mapping's src and dest differ,
// size-first. Returns false when the mapping is not a file-to-file pair.
func (m *Mapping) FileChanged() (bool, error) {
	if !m.FilesPresent() {
		return false, nil
	}
	return fsutil.FilesDiffer(m.sanitizedSrc, m.sanitizedDest)
}

func (m *Mapping) BackupPossible() bool { return m.Valid() && exists(m.sanitizedDest) }

// BackupBasename returns the name used for this mapping inside a backup dir, or
// "" when invalid. Mirrors backup_basename.
func (m *Mapping) BackupBasename() (string, bool) {
	if !m.Valid() {
		return "", false
	}
	if !exists(m.sanitizedDest) {
		return filepath.Dir(m.sanitizedDest), true
	}
	return filepath.Base(m.sanitizedDest), true
}

// --- path matching ---

// Include reports whether a path is inside an inclusion (used for file
// filtering). With no inclusions everything is included.
func (m *Mapping) Include(path string) bool {
	if !m.HasInclusions() {
		return true
	}
	if path == m.sanitizedSrc {
		return true
	}
	for _, incl := range m.sanitizedOnly {
		if inclusionMatches(incl, path) {
			return true
		}
	}
	return false
}

// BidirectionalInclude reports whether a path is inside OR an ancestor of an
// inclusion, letting directory traversal descend into parents of included
// files.
func (m *Mapping) BidirectionalInclude(path string) bool {
	if !m.HasInclusions() {
		return true
	}
	if path == m.sanitizedSrc {
		return true
	}
	for _, incl := range m.sanitizedOnly {
		if inclusionMatches(incl, path) || inclusionIsAncestor(path, incl) {
			return true
		}
	}
	return false
}

// Ignore reports whether path is ignored. NOTE: this reproduces Ruby's
// Mapping#ignore? which uses a raw String#start_with? (NOT a path-boundary
// check) — so an ignore entry "/a/foo" also matches "/a/foobar". The
// boundary-aware ignore filtering lives separately in the diff engine.
func (m *Mapping) Ignore(path string) bool {
	for _, ig := range m.sanitizedIgnores {
		if strings.HasPrefix(path, ig) {
			return true
		}
	}
	return false
}

func (m *Mapping) Skip(path string) bool { return m.Ignore(path) || !m.Include(path) }

// ShouldPruneDirectory reports whether an entire subtree can be skipped during
// traversal (the Find.prune optimization).
func (m *Mapping) ShouldPruneDirectory(path string) bool {
	if m.Ignore(path) {
		return true
	}
	if !m.HasInclusions() {
		return false
	}
	return !m.BidirectionalInclude(path)
}

func inclusionMatches(inclusion, path string) bool {
	if hasGlobMeta(inclusion) {
		return fnmatch(inclusion, path)
	}
	return paths.PathIsParentOrSame(inclusion, path)
}

func inclusionIsAncestor(path, inclusion string) bool {
	if hasGlobMeta(inclusion) {
		return paths.PathIsParentOrSame(path, filepath.Dir(inclusion))
	}
	return paths.PathIsParentOrSame(path, inclusion)
}

// ApplyTo returns a new single-file mapping for a changed path under src. Used
// by the watch daemon. Mirrors apply_to, including its dropping of inclusions
// (the Ruby original passes @only, which is never assigned and is therefore
// nil).
func (m *Mapping) ApplyTo(path string) *Mapping {
	rel := path
	if filepath.IsAbs(path) {
		rel = strings.TrimPrefix(path, m.sanitizedSrc+string(filepath.Separator))
	}
	return New(Attributes{
		Src:    filepath.Join(m.originalSrc, rel),
		Dest:   filepath.Join(m.originalDest, rel),
		Force:  m.force,
		Ignore: m.originalIgnores,
	})
}

// --- presentation ---

func (m *Mapping) DecoratedSrc() string  { return paths.ColorizeEnvVars(m.originalSrc) }
func (m *Mapping) DecoratedDest() string { return paths.ColorizeEnvVars(m.originalDest) }

// --- filesystem helpers (follow symlinks, matching Ruby File.* predicates) ---

func isDir(p string) bool {
	st, err := os.Stat(p)
	return err == nil && st.IsDir()
}

func isFile(p string) bool {
	st, err := os.Stat(p)
	return err == nil && st.Mode().IsRegular()
}

func exists(p string) bool {
	_, err := os.Stat(p)
	return err == nil
}
